tensorboard/plugins/graph/README.md

```diff
@@ -121,9 +121,384 @@ describing the motivations, research, and process behind it.
+
+        ```
+        {
+          ...node_def,
+          name: '__function_library__' + func.signature.name + '/' + node_def.name,
+          input: [node_def.input],
+          functionOutputIndex: currentOutputIndex++,
+        }
+        ```
+
+-   'Building the data structure' (SlimGraph)
+
+    -   Add nodes to SlimGraph
+
+        -   Set the name to be the normalized name.
+        -   Set the `inEmbeddings`, `outEmbeddings` properties on the OpNode.
+
+    -   Add edges to SlimGraph
+
+        -   Detect reference edges using a hardcoded list (e.g. Assign,
+            AssignAdd).
+        -   Move the inputs of the (Const) in-embedding node into incoming edges
+            of the main node.
+        -   Move the inputs of the (*Summary) out-embedding node into inputs of
+            the main node where the out-embedding points to.
+
+    -   Returns
+
+    ```
+    {
+      nodes: {
+        "adder/Add": OpNode,
+        ...
+      };
+      edges: BaseEdge[],
+      ...
+    }
+    ```
+
+### Building the Hierarchy
+
+View components create the RenderGraph mainly from the Hierarchy. As we will
+see, the key difference between a Hierarchy and a SlimGraph is that a Hierarchy
+contains more nodes beyond the OpNodes: MetaNodes are created to group other
+nodes, and MetaEdges are a layer on top of the SlimGraph's BaseEdges.
+
+Furthermore, the Hierarchy building phase will compute expensive information
+about the graph and store it on the data structure, such as templates. Template
+ids are stored on MetaNodes, and are used by the view components to determine
+node coloring. When 'Color by structure' is enabled, MetaNodes with similar
+structure are assumed to have the same templateId, and thus assigned the same
+color.
+
+The BridgeGraph is not computed at this phase, but instead on-demand by the view
+components.
+
+Note: building a Hierarchy modifies the SlimGraph's OpNodes in place!
+
+**tf_graph_common/hierarchy.ts: build(SlimGraph): Hierarchy**
+
+-   Create an artificial MetaNode for the root.
+-   Create an index of nodes as `Record<nodeId: string,
+    OpNode|SeriesNode|MetaNode>`
+-   'Adding nodes'
+    -   Iterate over OpNodes, keep track of device names, XLAClusters,
+        opHistograms, TPU Compatibility histograms, library function nodes and
+        their associated function name, etc.
+    -   Assume the node's name's '/' occurences indicate its level from the
+        root. If node is named 'a/b/c', create metanodes 'a' and 'a/b'.
+    -   Update embedding nodes to point to the OpNode they are embedded to:
+        `embedding.parentNode = node;`.
+-   'Detect series'
+    -   Recursively traverse metanodes starting from the top-level. At each
+        metanode,
+        -   Take the direct children nodes and group them into clusters by
+            op-type.
+        -   For each op cluster, detect series by greedily looking for numeric
+            suffix changes. For example, 3 OpNodes with op 'Add': `add_1`,
```

Comment:
It might be interesting to explicitly call out implicit `_0` which this logic understands.

Reply:
Done.